File: chazm-model/src/main/java/io/github/runtimemodels/chazm/entity/InstanceGoalEntity.java
package io.github.runtimemodels.chazm.entity;

import com.google.inject.assistedinject.Assisted;
import io.github.runtimemodels.message.M;
import lombok.Getter;
import runtimemodels.chazm.api.entity.InstanceGoal;
import runtimemodels.chazm.api.entity.SpecificationGoal;
import runtimemodels.chazm.api.id.UniqueId;

import javax.inject.Inject;
import javax.validation.constraints.NotNull;
import java.util.Objects;

class InstanceGoalEntity extends AbstractEntity<InstanceGoal> implements InstanceGoal {

    @Getter
    private final SpecificationGoal goal;
    @Getter
    private final InstanceGoal.Parameter parameter;
    private transient Integer hashCode = null;

    @Inject
    InstanceGoalEntity(@NotNull @Assisted final UniqueId<InstanceGoal> id, @NotNull @Assisted final SpecificationGoal goal,
                       @NotNull @Assisted final Parameter parameter) {
        super(id);
        this.goal = goal;
        this.parameter = parameter;
    }

    @Override
    public boolean equals(final Object object) {
        if (object instanceof InstanceGoal) {
            final InstanceGoal goal = (InstanceGoal) object;
            return super.equals(goal) && getGoal().equals(goal.getGoal());
        }
        return false;
    }

    @Override
    public int hashCode() {
        if (hashCode == null) {
            hashCode = Objects.hash(super.hashCode(), getGoal());
        }
        return hashCode;
    }

    @Override
    public String toString() {
        return M.ENTITY_2.get(super.toString(), getGoal(), getParameter());
    }

}

File: chazm-model/src/main/kotlin/runtimemodels/chazm/model/entity/InstanceGoalEntity.kt
package runtimemodels.chazm.model.entity

import com.google.inject.assistedinject.Assisted
import runtimemodels.chazm.api.entity.InstanceGoal
import runtimemodels.chazm.api.entity.SpecificationGoal
import runtimemodels.chazm.api.id.InstanceGoalId
import javax.inject.Inject

internal data class InstanceGoalEntity @Inject constructor(
    @Assisted override val id: InstanceGoalId,
    @param:Assisted override val goal: SpecificationGoal
) : InstanceGoal {
    internal data class DefaultGoalParameter(
        private val map: MutableMap<Any, Any> = mutableMapOf()
    ) : InstanceGoal.GoalParameter, Map<Any, Any> by map {
        override fun add(key: Any, value: Any) {
            map[key] = value
        }

        override fun remove(key: Any) {
            map.remove(key)
        }
    }

    override val parameters: InstanceGoal.GoalParameter = DefaultGoalParameter()
}
